include/librfu.h

```diff
@@ -93,143 +101,204 @@ struct __attribute__((packed, aligned(2))) UnkLinkRfuStruct_02022B14
     u8 unk_0a_0:7;
     u8 unk_0a_7:1;
     u8 playerGender:1;
-    u8 unk_0b_1:7;
+    u8 level:7;
     u8 unk_0c;
 };
 
 struct RfuStruct
 {
-    vs32 unk_0;
-    u8 txParams;
-    u8 unk_5;
-    u8 activeCommand;
-    u8 unk_7;
-    u8 unk_8;
-    u8 unk_9;
+    vs32 state;
+    u8 reqLength;
+    u8 reqNext;
+    u8 reqActiveCommand;
+    u8 ackLength;
+    u8 ackNext;
+    u8 ackActiveCommand;
     u8 timerSelect;
     u8 unk_b;
-    int timerState;
+    u32 timerState; // this field is s32 in emerald
     vu8 timerActive;
     u8 unk_11;
-    vu16 unk_12;
+    vu16 error;
     vu8 msMode;
-    u8 unk_15;
+    u8 recoveryCount;
     u8 unk_16;
     u8 unk_17;
     void (*callbackM)();
-    void (*callbackS)();
-    u32 callbackID;
-    union RfuPacket * txPacket;
-    union RfuPacket * rxPacket;
+    void (*callbackS)(u16);
+    void (*unk_20)(void);
+    union RfuPacket *txPacket;
+    union RfuPacket *rxPacket;
     vu8 unk_2c;
-    u8 padding[3];
 };
 
 struct RfuIntrStruct
 {
-    u8 rxPacketAlloc[0x74];
-    u8 txPacketAlloc[0x74];
-    u8 block1[0x960];
-    u8 block2[0x30];
+    union RfuPacket rxPacketAlloc;
+    union RfuPacket txPacketAlloc;
+    u8 block1[0x960]; // size of librfu_intr.s binary
+    struct RfuStruct block2;
 };
 
-struct RfuUnk1
+struct RfuSlotStatusUNI
 {
-    u16 unk_0;
-    u8 unk_2;
-    u8 unk_3;
-    u8 fill_4[14];
-    u8 unk_12;
-    u32 unk_14;
-    u32 unk_18;
-    struct RfuIntrStruct unk_1c;
+    struct
+    {
+        struct UNISend
+        {
+            u16 state;
+            u8 dataReadyFlag;
+            u8 bmSlot;
+            u16 payloadSize;
+            const void *src;
+        } send;
+        struct UNIRecv
+        {
+            u16 state;
+            u16 errorCode;
+            u16 dataSize;
+            u8 newDataFlag;
+            u8 dataBlockFlag;
+        } recv;
+    } sub;
+    void *recvBuffer;
+    u32 recvBufferSize;
 };
 
-struct RfuUnk2
+struct NIComm
 {
-    u16 unk_0;
-    u16 unk_2;
-    u8 fill_4[0x16];
-    u8 unk_1a;
-    u8 fill_1b[0x19];
-    u16 unk_34;
-    u16 unk_36;
-    u8 fill_38[0x16];
-    u8 unk_4e;
-    u8 fill_4f[0x12];
-    u8 unk_61;
-    u8 fill_62[6];
-    void *unk_68;
-    void *unk_6c;
-    u8 unk_70[0x70];
+    u16 state;
+    u16 failCounter;
+    const u8 *nowP[4];
+    s32 remainSize;
+    u16 errorCode;
+    u8 bmSlot;
+    u8 recvAckFlag[4];
+    u8 ack;
+    u8 phase;
+    u8 n[4];
+    const void *src;
+    u8 bmSlotOrg;
+    u8 dataType;
+    u16 payloadSize;
+    u32 dataSize;
 };
```

Comment:
```c
typedef struct NI_COMM_Tag {
    // Parameters indicating current communication state
    u16 state;                                             // Communication state of slot
    
    u16 failCounter;                                       // Count of failed transmissions/receptions (Count is increased when transmission/reception of data does not succeed within 1PF=16.7 ms)
    u8  *nowp[WINDOW_COUNT];                               // Address of current send/receive (The data is divided into WINDOW_COUNT blocks and sent in payloadSize units.)
    u32 remainSize;                                        // Size of remaining communication data
    u16 errorCode;                                         // Error code
    u8  bmSlot;                                            // Expresses the current communication slot in bits
                                                           // (When sending from the Maser, because multiple slaves can be specified with bmSlot, communications are terminated based on the failCounter for each child device)
    
    // Parameters used inside the Library
    u8  recv_ack_flag[WINDOW_COUNT];
    u8  ack;
    u8  phase;
    u8  n[WINDOW_COUNT];
    
    // Parameters indicating the current communication content
    const u8 *src;                                         // Start address transmission data (valid only for transmitting device)
    u8  bmSlot_org;                                        // Expresses the communication source slot at the beginning of the transmission in bits (valid only for transmitting device)
    u8  dataType;                                          // Transmission/reception data type (0: User data; 1: Game identification information)
    u16 payloadSize;                                       // Payload size for one transmission
    u32 dataSize;                                          // Size of transmission/reception data
                                                           // Calculated by subtracting link layer header size from subFrameSize specified by the rfu_NI_setSendData function    
}NI_COMM;
```